### ISSUE TITLE
Test for docker build success

### ISSUE THREAD

**Issue by rarkins** [renovateapp/renovate#1679]
The `node` `alpine` Docker base image we use "broke" yarn for us a few days ago. I had pinned the digest but I blindly accepted the Renovate PR which updated it. After all, what's the chances that a digest change to an existing Docker image tag by an org the size of `node` would break? Actually it did, and it's a perfect case study of why to pin Dockerfile images to their hashes always - at least now I can roll it back.

Anyway, ideally all PRs should be tested that they haven't broken our Docker image again, however this could also add quite a lot of time to *every* PR build on Travis if not done correctly.

Cc @adam-moss @felixfbecker @hulkish - might be of interest to you

**Comment by adam-moss** [comment]
🤦‍♂️ it was this issue https://github.com/nodejs/docker-node/issues/649

So actually for your purposes you can probably do something fairly similar to what I've been doing recently using [Google's Container Structure Test](https://github.com/GoogleCloudPlatform/container-structure-test) tool, with some test `.json` files like:

**/tests/container/file-existence.json**
```json
{
    "schemaVersion": "2.0.0",
    "fileExistenceTests": [{
        "name": "yarn",
        "path": "/usr/local/bin/yarn",
        "shouldExist": true,
        "permissions": "-rwxr-xr-x"
    }]
}
```

**/tests/container/commands.json**
```json
{
    "schemaVersion": "2.0.0",
    "commandTests": [{
        "name": "user account",
        "command": "whoami",
        "expectedOutput": [
            "node"
        ]
    },
    {
        "name": "renovate installation",
        "command": "node",
        "args": [
            "/src/lib/renovate.js",
            "--version"
        ],
        "expectedOutput": [
            "0.0.0-semantic-release"
        ]
    }]
}
```

Or you could use something like InSpec if you have a wider set of infra tests you want to run in relation to it.

I'm happy to work up an MR for this if you like?

**Comment by rarkins** [comment]
Could you explain more on how/when this would run? I think making sure it builds and has the correct user and access to npm and yarn are the bare minimum. But I’m already worried about how much wasted build minutes we’d consume considering how inefficient it can be to build images on CI sometimes, and the high percentage of PRs that don’t actually touch anything related to Docker

**Comment by adam-moss** [comment]
So the way I use it (in GitLab CI) is similar to this (although mine is broken up into stages as has clair scanning etc. also):

```yaml
        - docker build --label "org.label-schema.build-date=$(date +%Y-%m-%dT%T%z)" --label "org.label-schema.version=$CI_COMMIT_REF_NAME" --tag $BUILD_IMAGE .
        - docker push $BUILD_IMAGE
        - container-structure-test -test.v -image $BUILD_IMAGE $CI_PROJECT_DIR/tests/container/*.json
        - docker tag $BUILD_IMAGE $LATEST_IMAGE
        - docker push $LATEST_IMAGE
```

For your purposes rather than pushing to the public register (I'm using an internal one) we could probably `docker save` the tarball and pass that into Google's tool, rather than pushing and pulling an image.

I'm not overly familiar with Travis CI but hopefully we could make it smart enough to recognise what is/isn't appropriate per _and the high percentage of PRs that don’t actually touch anything related to Docker_ - it would be useful to understand what you mean here since obviously any code change has a material impact since it is included in the image.

**Comment by rarkins** [comment]
Actually, I have changed my mind about this for now. I would have caught the alpine node problem almost immediately if Docker Hub build failure notifications had been on in my account, and then rolled back immediately. I have enabled these notifications now and will leave testing as-is unless someone feels really strongly. I think this type of break should be pretty rare and our Dockerfile is very simple.
